src/output.h

```diff
@@ -25,39 +26,43 @@
 #ifndef _OUTPUT_H
 #define _OUTPUT_H
 
+#include <set>
+#include <string>
+
 #include <glib.h>
 #include "song-meta-data.h"
 
-// Feedback for the controlling part what is happening with the
-// output.
-enum PlayFeedback {
-  PLAY_STOPPED,
-  PLAY_STARTED_NEXT_STREAM,
-};
-typedef void (*output_transition_cb_t)(enum PlayFeedback);
-
-// In case the stream gets to know details about the song, this is a
-// callback with changes we send back to the controlling layer.
-typedef void (*output_update_meta_cb_t)(const struct SongMetaData *);
-
-int output_init(const char *shortname);
-int output_add_options(GOptionContext *ctx);
-void output_dump_modules(void);
-
-int output_loop(void);
-
-void output_set_uri(const char *uri, output_update_meta_cb_t meta_info);
-void output_set_next_uri(const char *uri);
-
-int output_play(output_transition_cb_t done_callback);
-int output_stop(void);
-int output_pause(void);
-int output_get_position(gint64 *track_dur_nanos, gint64 *track_pos_nanos);
-int output_seek(gint64 position_nanos);
-
-int output_get_volume(float *v);
-int output_set_volume(float v);
-int output_get_mute(int *m);
-int output_set_mute(int m);
+namespace Output {
+typedef enum OutputState { kPlaybackStopped, kStartedNextStream } OutputState;
+
+// Callbacks types from output to higher levels
+typedef void (*PlaybackCallback)(OutputState);
+typedef void (*MetadataCallback)(const TrackMetadata&);
+
+typedef std::set<std::string> MimeTypeSet;
+
+int AddOptions(GOptionContext* ctx);
+void DumpModules(void);
+int Loop(void);
+
+int Init(const char* shortname, PlaybackCallback play_callback,
+         MetadataCallback metadata_callback);
+
+MimeTypeSet GetSupportedMedia(void);
+
+void SetUri(const char* uri);
+void SetNextUri(const char* uri);
+
+int Play(void);
+int Pause(void);
+int Stop(void);
+int Seek(int64_t position_nanos);
+
+int GetPosition(int64_t& duration, int64_t& position);
```

Comment:
The old code was essentially using this bunch of functions to displatch to the right output module.
Longer term, we should consider to remove this artifact but instead pass around some object of type`OutputModule`. It will help enforcing a readable initialization sequence (construct the thing first, then pass it around).

For initaizttion, we can have one `static OutputModule* OutputModule::Create(...)`  factory that does what `Output::Init()` does now, but more explicit.

The whole `Loop()` stuff should also of course move out the output module.

But one thing at a time, these suggestions are merely a TODO for follow-up changes.